internal/config/config.go

```diff
@@ -0,0 +1,198 @@
+package config
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/fsnotify/fsnotify"
+	log "github.com/sirupsen/logrus"
+	"sigs.k8s.io/yaml"
+)
+
+// validateDistributions checks that the condition that
+// two distributions with the same distribution ID MUST NOT share the same prefix.
+// or in other terms, every pair of id,prefix (Distribution) must be unique
+func validateDistributions(distributions distributionsMap) error {
+	uniqueMap := make(map[string]struct{})
+
+	for _, value := range distributions {
+		hash := value.hashKey()
+		if _, ok := uniqueMap[hash]; ok {
+			return fmt.Errorf("error parsing configuration: distribution value duplicated id: %s prefix: %s", value.ID, value.Prefix)
+		}
+
+		uniqueMap[hash] = struct{}{}
+	}
+
+	return nil
+}
+
+func validateEntitlements(entitlements entitlementsMap, distributions distributionsMap) error {
+	entitlementDistros := make(map[string]string)
+
+	// build lookup map, used also to deduplicate distribution names
+	for eName, distros := range entitlements {
+		for _, distro := range distros {
+			entitlementDistros[distro] = eName
+		}
+	}
+
+	for dName, eName := range entitlementDistros {
+		if _, ok := distributions[dName]; !ok {
+			return fmt.Errorf("error parsing configuration: distribution %s in entitlement %s is not configured", dName, eName)
+		}
+	}
+
+	return nil
+}
+
+func (c *Config) parse(data []byte) error {
+	config := struct {
+		Distributions distributionsMap `json:"distributions"`
+		Entitlements  entitlementsMap  `json:"entitlements"`
+	}{}
+
+	if err := yaml.Unmarshal(data, &config); err != nil {
+		return err
+	}
+
+	err := validateDistributions(config.Distributions)
+	if err != nil {
+		return err
+	}
+
+	err = validateEntitlements(config.Entitlements, config.Distributions)
+	if err != nil {
+		return err
+	}
+
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	for name, value := range config.Distributions {
+		c.distributions[name] = value
+	}
+	for name := range c.distributions {
+		if _, ok := config.Distributions[name]; !ok {
+			delete(c.distributions, name)
+		}
+	}
+
+	for name, value := range config.Entitlements {
+		c.entitlements[name] = value
+	}
+	for name := range c.entitlements {
+		if _, ok := config.Entitlements[name]; !ok {
+			delete(c.entitlements, name)
+		}
+	}
```

Comment:
nit: we could just make new maps and set from config instead of adding and then deleting

```suggestion
        c.distributions = make(distributionsMap)	
	for name, value := range config.Distributions {
		c.distributions[name] = value
	}
	
	c.entitlements = make(entitlementsMap)	
	for name, value := range config.Entitlements {
		c.entitlements[name] = value
	}
```

Reply:
makes sense, much cleaner

Reply:
and with that the `New()` method can be removed too, that in general didn't make much sense in first place

Reply:
done